### PR TITLE
Add QueryOption to support custom Dialer

### PR DESCRIPTION
**Pull request by AlCutter**
This PR adds support for using a custom dialer when contacting the configured NTP server:
 - A new `Dial` function on the `QueryOptions` struct allows the custom dialer to be passed in, defaulting to the original built-in code using Go's `net` package when unset.
 - A new `TimeWithOptions` func, matching the `QueryWithOptions` func, to continue to support the simple "one-call" mode when using a custom dialer.

This functionality is important if the application wanting to use this library is dependant upon something like [gVisor](https://pkg.go.dev/gvisor.dev/gvisor/pkg/tcpip) to provide the networking stack. 

(In my case, I'm building a bare-metal unikernel based on [TamaGo](https://github.com/usbarmory/tamago), where there is no OS to enable Go's `net` package to work.)

### CODE REVIEW

**Comment by beevik** [comment]
Thank you for your submission.  I am happy to accept the change with a few modifications.  Please modify, squash your commits, and resubmit.  Thanks!

**Comment by AlCutter** [comment]
> Thank you for your submission. I am happy to accept the change with a few modifications. Please modify, squash your commits, and resubmit. Thanks!

I've squashed and pushed, were there any modifications you wanted me to make to the contents of the PR?

**Comment by beevik** [comment]
Yes, I submitted a bunch of comments in the review.  You should be able to see them above the conversation.

File: ntp.go [hunk 1]
```diff
@@ -167,6 +168,14 @@ type QueryOptions struct {
...
 	Port         int           // Server port, defaults to 123
 	TTL          int           // IP TTL to use, defaults to system default
+	// Dial allows custom dialer functions to be used when contacting the
+	// NTP server.
+	// Defaults to using Go's built-in UDP support fron the net package.
+	// Implementations should use rHostPort as the "host:port" of the NTP server
+	// to dial, and lHostPort can optionally be used to specify the "host:port"
+	// of the local socket.
+	// Should return a Conn representing a UDP connection to the server.
+	Dial func(lHostPort, rHostPort string) (net.Conn, error)
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply ntp.go]: Done x2

File: ntp.go [hunk 2]
```diff
@@ -436,6 +436,28 @@ func getTime(host string, opt QueryOptions) (*msg, ntpTime, error) {
...
+	raddr, err := net.ResolveUDPAddr("udp", rHostPort)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve rHostPort %q: %v", rHostPort, err)
+	}
+
+	var laddr *net.UDPAddr
+	if lHostPort != "" {
+		laddr, err = net.ResolveUDPAddr("udp", lHostPort)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve lHostPort %q: %v", lHostPort, err)
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply ntp.go]: ditto

File: README.md [hunk 3]
```diff
@@ -69,3 +69,10 @@ To prevent it from becoming overloaded, please avoid querying the standard
 `pool.ntp.org` zone names in your applications.  Instead, consider requesting
 your own [vendor zone](http://www.pool.ntp.org/en/vendors.html) or [joining
 the pool](http://www.pool.ntp.org/join.html).
+
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply README.md]: Reverted

**Review by beevik — APPROVED** [review]
looks good, thanks for making the changes.

**Pull request merged**